server/src/main/java/org/opensearch/common/remote/RemoteWritableEntityManager.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.remote;
+
+import org.opensearch.common.CheckedRunnable;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.gateway.remote.ClusterMetadataManifest.UploadedMetadata;
+import org.opensearch.gateway.remote.model.RemoteReadResult;
+
+import java.io.IOException;
+
+/**
+ * The RemoteWritableEntityManager interface provides async read and write methods for managing remote entities in the remote store
+ */
+public interface RemoteWritableEntityManager {
+
+    /**
+     * Returns a CheckedRunnable that performs an asynchronous read operation for the specified component and entity.
+     *
+     * @param component the component for which the read operation is performed
+     * @param entity the entity to be read
+     * @param listener the listener to be notified when the read operation completes
+     * @return a CheckedRunnable that performs the asynchronous read operation
+     */
+    CheckedRunnable<IOException> asyncRead(
+        String component,
```

Comment:
why is component needed?

Reply:
this is needed because, we have cases where `entity.getType()` not return complete detail of the entity we are reading/writing. For ex, if we are reading/writing a index metadata `entity.getType()` will be just INDEX but we need to distinguish it by adding index name/uuid as well. Similarly with customs, we need to know the custom type we are uploading. This is populated in RemoteReadResult when we are sending the response to the listener. This is not necessarily required in write, but still added this to make it consistent.